src/prefect/orion/orchestration/core_policy.py

```diff
@@ -450,3 +465,137 @@ async def before_transition(
             )
+
+
+class OnlyRestartFromTerminalStates(BaseOrchestrationRule):
+    FROM_STATES = set(ALL_ORCHESTRATION_STATES) - set(TERMINAL_STATES)
+    TO_STATES = [states.StateType.SCHEDULED]
+
+    async def before_transition(
+        self,
+        initial_state: Optional[states.State],
+        proposed_state: Optional[states.State],
+        context: OrchestrationContext,
+    ) -> None:
+        await self.abort_transition(
+            reason="Can only restart runs that have terminated."
+        )
+
+
+class PreventRestartingSubflowRuns(BaseOrchestrationRule):
+    FROM_STATES = ALL_ORCHESTRATION_STATES
+    TO_STATES = ALL_ORCHESTRATION_STATES
+
+    async def before_transition(
+        self,
+        initial_state: Optional[states.State],
+        proposed_state: Optional[states.State],
+        context: OrchestrationContext,
+    ) -> None:
+        if context.run.parent_task_run_id:
+            await self.abort_transition("Cannot restart a subflow run.")
+
+
+class PreventRestartingFlowsWithoutDeployments(BaseOrchestrationRule):
+    FROM_STATES = ALL_ORCHESTRATION_STATES
+    TO_STATES = ALL_ORCHESTRATION_STATES
+
+    async def before_transition(
+        self,
+        initial_state: Optional[states.State],
+        proposed_state: Optional[states.State],
+        context: OrchestrationContext,
+    ) -> None:
+        if not context.run.deployment_id:
+            await self.abort_transition(
+                "Cannot restart a run without an associated deployment."
+            )
+
+
+class RestartFlowRun(BaseOrchestrationRule):
+    FROM_STATES = TERMINAL_STATES
+    TO_STATES = [states.StateType.SCHEDULED]
+
+    async def before_transition(
+        self,
+        initial_state: Optional[states.State],
+        proposed_state: Optional[states.State],
+        context: OrchestrationContext,
+    ):
+        # update run count
+        self.original_run_count = context.run.run_count
+        context.run.run_count = 0  # reset run count to preserve retry behavior
+
+        # update empirical settings
+        self.original_settings = context.run_settings.copy()
+        self.restarts = self.original_settings.restarts + 1
+        new_settings = context.run_settings.copy()
+        new_settings.restarts = self.restarts
+        flow_run_update = actions.FlowRunUpdate(empirical_policy=new_settings)
+        await flow_runs.update_flow_run(
+            context.session, context.run.id, flow_run_update
+        )
+
+    async def after_transition(
+        self,
+        initial_state: Optional[states.State],
+        proposed_state: Optional[states.State],
+        context: OrchestrationContext,
+    ):
+        from prefect.orion.models import task_runs
+
+        trs = await flow_runs.read_task_runs(context.session, context.run.id)
+        for task_run in trs:
+            if task_run.empirical_policy.flow_restart_index is None:
```

Comment:
Could you explain this logic?

Reply:
The hard restart strategy keeps all the old task runs and states (also important to caching), but it's important to indicate that the the task runs came from a different "attempt" at traversing the flow logic. If a user presses the restart button twice on a flow with 5 tasks, the flow run will end up with 15 tasks runs which we can distinguish in the UI with this metadata

![image](https://user-images.githubusercontent.com/1688064/195666961-38234dd0-9a6c-4a9f-8873-b20a590c337b.png)

Reply:
Hm that doesn't sound right to me - we shouldn't create any new task runs with a flow run restart; I also don't see where any new task runs are created in the PR?  But tracking what attempt a task run was successful on makes sense to me

Reply:
if we don't make new task runs the restart will always fail to work properly if there was a change in execution environment that would change the "realized task graph" of the flow run, given that this graph can be different from restart to restart (e.g. if I misconfigured a block or I forget to turn on a server--anything outside of Prefect's control) it's very difficult to the tasks to each other properly

Reply:
the idea here is: if users have configured their cache keys and result storage properly, the restarted task runs never enter a running state, they are retrieved from the cached state created from the original attempt instead.

We do work to try and match up "old runs" with "new runs" in our flow retry logic. It relies on idempotency key and will break even if the same graph is realized in a different order—something as simple as a task taking longer to run when the target bucket is filled with data can mess up the task run matching; I do think if we really want to try and match up task runs there are options there to do so based on the cache key instead

Reply:
sorry--to answer your question about where the task runs come from: the flow run is put in a scheduled state, so an agent will pick up the flow again and create new task runs as it runs the flow

Reply:
Two points:
- putting the flow run into a scheduled state won't create any new task runs (and that's a good thing)
- it's OK for us to tell users they need to write stable graphs to maximally benefit from our feature set; creating new runs is going to be very noisy and not representative of what's happening (in reality the runs are the same, they just have a longer state history)

Reply:
We can't loop over every task run and modifying it when setting a flow run state - this is the exact pattern that frequently kills Prefect 1 deployments with large number of task runs (imagine mapping over thousands of tasks and trying to retry!)

I suspect that the information about what "flow run iteration" a task run belongs to can be inferred by their relative timestamps if necessary, but if we want to encode it at the task run level then the time to do so is when the task run state is recorded. E.g. when marking a task run as complete, also record the flow run's retry index at that time. This will achieve the same outcome with no additional overhead.

Reply:
We are already doing this for automated flow run retries. Perhaps we need to take a step back and take a look at our mechanisms for performing a rerun holistically?